new changes 12/19

alxqiu opened the pull request:
defined parameters for interface, including host and port info, as well as leaving optional 'path' and 'parameters' for initialization of executor class

Commit:
new changes 4:46pm 11/14/2019

Commit:
new changes for interface, including tested POST and GET requests.

ningoneclick commented on Interface.py:
```diff
@@ -8,56 +8,60 @@
-        #server stores info of jobs in centralized location, so should manage encoding
+        print(self.url_str)
 
     def submit(self, fn, job_name, memory, *args, priority = 1, retries = 0, run_now = False, **kwargs):
```
please use the same signature as the submit method in the concurrent lib. The additional job-specific policy parameters can be defined in a separate parameter, called job_config, which you can define as dict

ningoneclick commented on Interface.py:
```diff
@@ -8,56 +8,60 @@
-        #can include map() later not now.  
-        
     #next I want to include a function that will grab info from the server
-    def grab(self, job_name):
```
Internal methods should start with underscore "_"

ningoneclick commented on Interface.py:
```diff
@@ -75,19 +79,33 @@ def shutdown(self, wait = True):
 #and should keep all the info for the server and name, and facilitate communication between server and client
 class Future():
-    def __init__(self, function_name, server_info, func_args, func_kwargs):
+    def __init__(self, function_name, server_info, memory, priority, retries, *args, **kwargs):
```
keep argument's name consistent with variable name.
instead of server_info, keep a reference to the Executor object.
I don't think you need any other parameters here.

Commit:
added new parameter for the submit function to mirror format of
concurrent.futures.Executor.submit()

alxqiu committed:
Add files via upload

###This is a duplicate of what I committed to Master###
TestScript.py : This tests get and post requests sent by Interface.py, and prints out the responses from the "Server"
ServerSideScript.py :  This facilitates basic responses as a simple "Server"

Commit:
New changes, but functional for web server under newWebServerStuff/

Commit:
Updated functional interface working with server

Commit:
'Change'

Commit:
New changes to the branch after it seemed to work on github

 Branch 'newbranch11/13' of https://github.com/alxqiu/Scheduler into newbranch11/13

Commit:
New commits 12/20

Commit:
New commits 12/23

Commit:
New commits for readability

Commit:
New changes to future objects

Commit:
New changes for 1/7

Commit:
New changes to job_manager

Commit:
Finished queue of running jobs

ningoneclick commented on Interface.py:
```diff
@@ -1,72 +1,195 @@
+            time.sleep(1.0)
+            timer += 1.0
+            r = requests.get(url = self.url_str + '/grab-job-info', json = {'job_id': self.job_id})
+            if (r.json['cancelled']):
```
you perhaps better store the status in a single property, instead of using multiple booleans. Those status are mutually exclusive

ningoneclick commented on Interface.py:
```diff
@@ -1,72 +1,195 @@
+            r = requests.get(url = self.url_str + '/grab-job-info', json = {'job_id': self.job_id})
+            if (r.json['cancelled']):
+                raise CancelledError
+            if (r.json['exception'] is not None):
```
Do you mean `elif`

ningoneclick commented on Interface.py:
```diff
@@ -1,72 +1,195 @@
+                raise CancelledError
+            if (r.json['exception'] is not None):
+                #develop system to have the same exceptions
+                raise r.json['exception']
```
The exception is an serialized object from server. You should deserialize it into object and then raise it up.

ningoneclick commented on Interface.py:
```diff
@@ -1,72 +1,195 @@
+            if (timer > timeout):
+                raise TimeoutError
+        #per project parameters, this will return None if no exception is there
+        return r.json()['exception']
```
Similar comments here

ningoneclick commented on Interface.py:
```diff
@@ -1,72 +1,195 @@
+        #as they take the future object as the only parameter
+        #they can be as simple as printing the results of the job
+                ##########
+    def add_done_callback(self, fn):
```
may be call the call back function `callback`. It is less confusing

ningoneclick replied:
add_done_callback means to register a function to run after the job finishes. It should not block the client code. I will explian its implementation in person

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+import datetime
+from flask import Flask, request
+import json
+from http.server import BaseHTTPRequestHandler, HTTPServer
```
Remove if not used

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+    return 'Up and running!'
+
+#sends back unique id for a function in str form so its static
+def function_encoder(fn):
```
Use uuid to avoid conflicts

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+        #every iteration of this large loop is one job ran
+        #adjust sleep time arbitrarily, or by system constraints
+        time.sleep(2.4)
+        if (data_for_jobs is not None):
```
data_for_jobs is never None

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+        if (data_for_jobs is not None):
+            #update the viable jobs list
+            runnable_jobs = dict()        
+            for job in data_for_jobs.items():
```
Write it like
`for job_id, job in data_for_jobs.items()`

ningoneclick replied:
the access to data_for_jobs needs to be guarded by locks, to pervent simultaneous write access (which modify data_for_jobs)

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+            target_id = list(runnable_jobs.keys())[0]
+            highest_prior = runnable_jobs.get(list(runnable_jobs.keys())[0]).get('priority')
+            #now it can trump priority, and defers to oldest. 
+            for job in runnable_jobs.items():
```
access to job objects may also modify its property. Put inside a lock

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+    func_args = pickle.loads(data_for_jobs.get(job_id)['func_args'])
+    func_kwargs = pickle.loads(data_for_jobs.get(job_id)['func_kwargs'])
+    data_for_jobs.get(job_id)['running'] = True
+    for i in range(data_for_jobs.get(job_id)['retries'] + 1):
```
why do you use get?

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+    data_for_jobs.get(job_id)['running'] = True
+    for i in range(data_for_jobs.get(job_id)['retries'] + 1):
+         try:
+             fn(func_args, func_kwargs)
```
you need to capture its return value

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+            #####or do I if the running function is passively running
+
+    #the variable containing the dict of job information sent through is updated with an unique id,  
+    data_for_jobs[encoded_var] = request_data
```
locks

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+    request_data = request.get_json()
+
+    #adding encoded data
+    encoded_var = function_encoder(request_data['fn'])
```
do you mean decoded

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+
+    #the variable containing the dict of job information sent through is updated with an unique id,  
+    data_for_jobs[encoded_var] = request_data
+    return (request_data)
```
you are sending all the request parameters back too, which is not needed.

ningoneclick commented on ServerSideScript.py:
```diff
@@ -0,0 +1,170 @@
+@ServerSideScript.route('/grab-job-info', methods = ['GET'])
+def grab_job_info():
+    send = json.dumps(data_for_jobs.get(str(request.get_json()['job_id'])))
+    return send
```
locks

Commit:
Implemented pickle and multithreading, function serialization and deserialization should work now

Commit:
Fixed debug functions, but can't seem to make jobs run 2/5/2020

Commit:
attempt at implementing multithreading to solve jobs not running

Commit:
Just debugged results, now looking to debug non-primary functions

Commit:
Finalized Changes
